File: backend/src/services/pharmacyService.js
const prisma = require('../config/db');
const { logAudit } = require('./auditService');
const { createNotification } = require('./notificationService');
const { addInvoiceLineItem } = require('./billingService');

/**
 * Transactionally dispenses medicines, deducts stock, adds pharmacy invoice line items,
 * and raises low-stock alerts.
 */
async function dispenseMedicinesTransaction({
  patientId,
  items, // array of { medicineId, quantity }
  dispensedByUserId = null,
  ipAddress = null
}) {
  if (!items || items.length === 0) {
    throw new Error('At least one medicine item must be specified for dispensing');
  }

  return await prisma.$transaction(async (tx) => {
    const dispensedDetails = [];

    for (const item of items) {
      const { medicineId, quantity } = item;
      const qty = parseInt(quantity);

      if (qty <= 0) {
        throw new Error('Dispense quantity must be greater than 0');
      }

      // Fetch current medicine record
      const medicine = await tx.medicine.findUnique({
        where: { id: medicineId }
      });

      if (!medicine) {
        throw new Error(`Medicine with ID ${medicineId} not found`);
      }

      if (medicine.quantity < qty) {
        throw new Error(
          `Insufficient inventory for ${medicine.name}. Requested: ${qty}, Available: ${medicine.quantity}`
        );
      }

      const newQty = medicine.quantity - qty;

      // Atomically deduct inventory
      const updatedMedicine = await tx.medicine.update({
        where: { id: medicineId },
        data: { quantity: newQty }
      });

      // Add to pharmacy billing invoice
      const dispenseRefId = `DISP_${Date.now()}_${medicineId.slice(0, 6)}`;
      const { invoice } = await addInvoiceLineItem({
        patientId,
        billingType: 'PHARMACY',
        sourceDepartment: 'PHARMACY',
        itemBillingType: 'MEDICINE',
        itemDescription: `${medicine.name} (Qty: ${qty})`,
        quantity: qty,
        unitPrice: medicine.unitPrice,
        medicineId: medicine.id,
        sourceEntity: 'PHARMACY_SALE',
        sourceId: dispenseRefId,
        prismaClient: tx
      });

      dispensedDetails.push({
        medicine: updatedMedicine,
        dispensedQuantity: qty,
        unitPrice: medicine.unitPrice,
        totalCost: qty * medicine.unitPrice
      });

      // Low-stock & Out-of-stock notification triggers
      if (newQty === 0) {
        await createNotification({
          role: 'PHARMACIST',
          title: 'Critical: Medicine Out of Stock',
          message: `${medicine.name} (${medicine.code}) is now OUT OF STOCK (0 units remaining).`,
          type: 'OUT_OF_STOCK',
          entityId: medicine.id,
          prismaClient: tx
        });
        await createNotification({
          role: 'ADMIN',
          title: 'Inventory Alert: Stock Exhausted',
          message: `${medicine.name} (${medicine.code}) has reached 0 units.`,
          type: 'OUT_OF_STOCK',
          entityId: medicine.id,
          prismaClient: tx
        });
      } else if (newQty <= medicine.reorderThreshold) {
        await createNotification({
          role: 'PHARMACIST',
          title: 'Low Medicine Stock Alert',
          message: `${medicine.name} (${medicine.code}) stock level is down to ${newQty} units (Reorder threshold: ${medicine.reorderThreshold}).`,
          type: 'LOW_STOCK',
          entityId: medicine.id,
          prismaClient: tx
        });
        await createNotification({
          role: 'ADMIN',
          title: 'Pharmacy Reorder Threshold Breached',
          message: `${medicine.name} (${medicine.code}) has dropped to ${newQty} units.`,
          type: 'LOW_STOCK',
          entityId: medicine.id,
          prismaClient: tx
        });
      }
    }

    // Record audit trail
    await logAudit({
      userId: dispensedByUserId,
      action: 'PHARMACY_DISPENSE',
      resource: 'PHARMACY',
      details: {
        patientId,
        itemsDispensed: dispensedDetails.map(d => ({
          name: d.medicine.name,
          quantity: d.dispensedQuantity,
          remaining: d.medicine.quantity
        }))
      },
      ipAddress,
      prismaClient: tx
    });

    return { success: true, dispensed: dispensedDetails };
  });
}

module.exports = {
  dispenseMedicinesTransaction
};

File: backend/app/services/pharmacy_service.py
import time
from typing import Optional, List, Dict, Any
from sqlalchemy.orm import Session
from app import models
from app.services.billing_service import add_invoice_line_item
from app.services.notification_service import create_notification
from app.services.audit_service import log_audit

def dispense_medicines_transaction(
    db: Session,
    patient_id: str,
    items: List[Dict[str, Any]],
    dispensed_by_user_id: Optional[str] = None,
    ip_address: Optional[str] = None
) -> Dict[str, Any]:
    if not items:
        raise ValueError("At least one medicine item must be specified for dispensing")

    dispensed_details = []

    for itm in items:
        med_id = itm.get("medicineId") if isinstance(itm, dict) else getattr(itm, "medicineId")
        qty = int(itm.get("quantity") if isinstance(itm, dict) else getattr(itm, "quantity"))

        if qty <= 0:
            raise ValueError("Dispense quantity must be greater than 0")

        medicine = db.query(models.Medicine).filter(models.Medicine.id == med_id).first()
        if not medicine:
            raise ValueError(f"Medicine with ID {med_id} not found")

        if medicine.quantity < qty:
            raise ValueError(f"Insufficient inventory for {medicine.name}. Requested: {qty}, Available: {medicine.quantity}")

        new_qty = medicine.quantity - qty
        medicine.quantity = new_qty
        db.flush()

        dispense_ref_id = f"DISP_{int(time.time() * 1000)}_{med_id[:6]}"
        add_invoice_line_item(
            db=db,
            patient_id=patient_id,
            billing_type="PHARMACY",
            source_department="PHARMACY",
            item_billing_type="MEDICINE",
            item_description=f"{medicine.name} (Qty: {qty})",
            quantity=qty,
            unit_price=medicine.unitPrice,
            medicine_id=medicine.id,
            source_entity="PHARMACY_SALE",
            source_id=dispense_ref_id
        )

        dispensed_details.append({
            "medicine": medicine,
            "dispensedQuantity": qty,
            "unitPrice": medicine.unitPrice,
            "totalCost": qty * medicine.unitPrice
        })

        if new_qty == 0:
            create_notification(
                db=db,
                role="PHARMACIST",
                title="Critical: Medicine Out of Stock",
                message=f"{medicine.name} ({medicine.code}) is now OUT OF STOCK (0 units remaining).",
                type="OUT_OF_STOCK",
                entity_id=medicine.id
            )
            create_notification(
                db=db,
                role="ADMIN",
                title="Inventory Alert: Stock Exhausted",
                message=f"{medicine.name} ({medicine.code}) has reached 0 units.",
                type="OUT_OF_STOCK",
                entity_id=medicine.id
            )
        elif new_qty <= medicine.reorderThreshold:
            create_notification(
                db=db,
                role="PHARMACIST",
                title="Low Medicine Stock Alert",
                message=f"{medicine.name} ({medicine.code}) stock level is down to {new_qty} units (Reorder threshold: {medicine.reorderThreshold}).",
                type="LOW_STOCK",
                entity_id=medicine.id
            )
            create_notification(
                db=db,
                role="ADMIN",
                title="Pharmacy Reorder Threshold Breached",
                message=f"{medicine.name} ({medicine.code}) has dropped to {new_qty} units.",
                type="LOW_STOCK",
                entity_id=medicine.id
            )

    log_audit(
        db=db,
        user_id=dispensed_by_user_id,
        action="PHARMACY_DISPENSE",
        resource="PHARMACY",
        details={
            "patientId": patient_id,
            "itemsDispensed": [
                {
                    "name": d["medicine"].name,
                    "quantity": d["dispensedQuantity"],
                    "remaining": d["medicine"].quantity
                } for d in dispensed_details
            ]
        },
        ip_address=ip_address
    )

    db.commit()
    return {"success": True, "dispensed": dispensed_details}
